generator/sidekick/downloads_cache.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/walle/targz"
+)
+
+func makeGoogleapisRoot(rootConfig *Config) (string, error) {
+	googleapisRoot, ok := rootConfig.Source["googleapis-root"]
+	if !ok {
+		return "", nil
+	}
+	if ok := isDirectory(googleapisRoot); ok {
+		return googleapisRoot, nil
+	}
+	if !requiresDownload(googleapisRoot) {
+		return "", fmt.Errorf("only directories and https URLs are supported for googleapis-root")
+	}
+	// Treat `googleapis-root` as a URL to download. We want to avoid downloads
+	// if possible, so we will first try to use a cache directory in $HOME.
+	// Only if that fails we try a new download.
+	googleapisSha256, ok := rootConfig.Source["googleapis-sha256"]
+	if !ok {
+		return "", fmt.Errorf("using an https:// URL for googleapis-root requires setting googleapis-sha256")
+	}
+	cacheDir, err := downloadsCacheDir(rootConfig)
+	if err != nil {
+		return "", err
+	}
+	target := path.Join(cacheDir, googleapisSha256)
+	if isDirectory(target) {
+		return target, nil
+	}
+	tgz := target + ".tar.gz"
+	if err := downloadGoogleapisRoot(tgz, googleapisRoot, googleapisSha256); err != nil {
+		return "", err
+	}
+
+	if err := targz.Extract(tgz, cacheDir); err != nil {
+		return "", err
+	}
+	name := extractedName(rootConfig, googleapisRoot)
+	if err := os.Rename(path.Join(cacheDir, name), target); err != nil {
+		return "", err
+	}
+	return target, nil
+}
+
+func extractedName(rootConfig *Config, googleapisRoot string) string {
+	name, ok := rootConfig.Source["googleapis-extracted-name"]
+	if ok {
+		return name
+	}
+	return "googleapis-" + filepath.Base(strings.TrimSuffix(googleapisRoot, ".tar.gz"))
+}
+
+func downloadGoogleapisRoot(target, source, sha256 string) error {
+	if fileExists(target) {
+		return nil
+	}
+	var err error
+	for range 3 {
+		if err = downloadAttempt(target, source, sha256); err == nil {
```

Comment:
should there be an exponential backoff here?

Reply:
Good idea.